[CIR] Add emitDeclInvariant for global with constant storage

adams381 opened the pull request:
Implement emitDeclInvariant to emit llvm.invariant.start intrinsic for global variables with constant storage. This enables optimizations by marking when a global becomes read-only after initialization.

## Changes
- Add emitDeclInvariant and emitInvariantStart functions in CIRGenCXX.cpp
- Add emitInvariantStart declaration in CIRGenFunction.h
- Update emitCXXGlobalVarDeclInit to call emitDeclInvariant for constant storage globals after initialization
- Update getOrCreateCIRGlobal to set constant flag on globals with constant storage
- Add comprehensive test covering positive and negative cases

## Implementation Details
The implementation handles address spaces correctly, dynamically constructing the intrinsic name (e.g., invariant.start.p0, invariant.start.p10) based on the pointer's address space. The intrinsic is only emitted when optimizations are enabled, matching classic codegen behavior.

## Testing
All tests pass (411/412, 1 unsupported). The test file includes CIR, LLVM, and OGCG checks for both optimized and non-optimized builds.

adams381 committed:
[CIR] Add emitDeclInvariant for global with constant storage

Implement emitDeclInvariant to emit llvm.invariant.start intrinsic for
global variables with constant storage. This enables optimizations by
marking when a global becomes read-only after initialization.

Changes:
- Add emitDeclInvariant and emitInvariantStart functions in CIRGenCXX.cpp
- Add emitInvariantStart declaration in CIRGenFunction.h
- Update emitCXXGlobalVarDeclInit to call emitDeclInvariant for constant
  storage globals after initialization
- Update getOrCreateCIRGlobal to set constant flag on globals with
  constant storage
- Add comprehensive test covering positive and negative cases

The implementation handles address spaces correctly, dynamically
constructing the intrinsic name (e.g., invariant.start.p0,
invariant.start.p10) based on the pointer's address space.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -652,10 +652,19 @@ CIRGenModule::getOrCreateCIRGlobal(StringRef mangledName, mlir::Type ty,
 
   mlir::Location loc = getLoc(d->getSourceRange());
 
+  // Calculate constant storage flag before creating the global.
```
Why did you move this?

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenModule.cpp:
```diff
@@ -870,10 +875,14 @@ void CIRGenModule::emitGlobalVarDefinition(const clang::VarDecl *vd,
     emitter->finalize(gv);
 
   // If it is safe to mark the global 'constant', do so now.
+  // Use the same logic as emitCXXGlobalVarDeclInit to determine constant
```
This code does not use the same logic as emitCXXGlobalVarDeclInit in classic codegen. I don't know why it doesn't but it doesn't. This change could lead us to behave differently.

andykaylor commented on clang/test/CIR/CodeGen/global-constant-storage.cpp:
```diff
@@ -0,0 +1,244 @@
+const C c;
+
+// CIR checks for 'a' - should have constant storage
+// CIR: cir.global constant external @a = #cir.zero : !rec_A
```
The initial CIR generated for this case should look like this:

```
  cir.global external dso_local @a = ctor : !rec_A {
    %0 = cir.get_global @a : !cir.ptr<!rec_A>
    cir.call @_ZN1AC1Ev(%0) : (!cir.ptr<!rec_A>) -> ()
  }
```
That gets transformed into what you're checking here during LoweringPrepare, but I'd like to see checks for both states. You can do that by passing `-mmlir --mlir-print-ir-before=cir-lowering-prepare` on the RUN line and capturing the stderr output to a different file. Look at `clang/test/CIR/CodeGen/global-init.cpp` for an example.

adams381 committed:
[CIR] Fix parameter naming in emitDeclInvariant

Change parameter names from CGF/D to cgf/d to match LLVM coding
standards. Addresses review comments.

adams381 committed:
[CIR] Fix parameter naming in emitInvariantStart

Change parameter names from PascalCase (Size, Addr) to lowercase
(size, addr) to match LLVM coding standards for function parameters.

adams381 committed:
[CIR] Remove address space suffix from invariant.start intrinsic

Use base intrinsic name 'invariant.start' instead of constructing
address space-specific names (e.g., 'invariant.start.p0'). The address
space will be automatically handled when the intrinsic is lowered to
LLVM IR, matching classic codegen behavior.

Also add comment explaining why constant storage calculation was moved
to before global creation in CIRGenModule.cpp.

adams381 committed:
[CIR] Align emitGlobalVarDefinition constant logic with classic codegen

Match classic codegen EmitGlobalVarDefinition logic by checking
!needsGlobalCtor && !needsGlobalDtor before calling isConstantStorage.
This ensures CIR behaves consistently with classic codegen for global
variable constant flag determination.

The isConstantStorage call now uses ExcludeCtor=true and ExcludeDtor=true
to match classic codegen, instead of the previous logic that used
!needsDtor for ExcludeDtor.

Update test expectations to match correct behavior: globals that need
constructors are not marked as constant, even if they have constant
storage semantics (which still triggers invariant.start emission).

adams381 committed:
[CIR] Remove duplicate CWithDtor case from test

Remove CWithDtor test case as it duplicates the A case. The C case
already exists and tests the simple case with internal linkage, matching
classic codegen test patterns.

Update function numbering: C case changes from __cxx_global_var_init.4
to __cxx_global_var_init.3 after removing the duplicate case.

adams381 committed:
[CIR] Reorganize test checks to interleave by test case

Reorganize global-constant-storage.cpp test file to group CIR, LLVM, and
OGCG checks together for each test case, making it easier to compare
the three implementations side-by-side.

This addresses reviewer feedback requesting that checks be interleaved
so that the various implementations of each case are adjacent.

adams381 committed:
[CIR] Add checks for before/after LoweringPrepare transformation

Add CIR-BEFORE-LPP checks to verify the initial CIR format with ctor
regions before LoweringPrepare transformation. This shows globals in
their initial state (cir.global external @a = ctor : !rec_A { ... })
before they are transformed into separate init functions.

This addresses reviewer feedback requesting checks for both the initial
CIR state and the state after LoweringPrepare transformation.

adams381 committed:
[CIR] Add -O1 RUN line to test CIR output at optimization level

Add a RUN line to emit CIR at -O1 optimization level and add corresponding
FileCheck assertions to verify that invariant.start intrinsic calls are
present in the CIR representation for constant storage cases.

This addresses reviewer feedback requesting a RUN line to test CIR output
at -O1, ensuring that optimization level correctly triggers the emission
of invariant.start intrinsics in CIR.

adams381 commented:
All reviewer feedback has been addressed:

✅ Parameter naming - Changed to camelCase
✅ Address space handling - Removed dynamic suffix construction
✅ Comment explaining move - Added explanation for constant storage calculation move
✅ Logic consistency - Aligned emitGlobalVarDefinition with classic codegen
✅ Remove CWithDtor case - Removed duplicate test case
✅ Reorganize test checks - Interleaved checks by test case
✅ Add checks for before/after LoweringPrepare - Added CIR-BEFORE-LPP checks
✅ Add -O1 RUN line for CIR output - Added CIR-O1 checks

Ready for re-review.

andykaylor reviewed (approved):
lgtm
